exotica_py/src/Exotica.cpp

```diff
@@ -0,0 +1,508 @@
+    object.def("getType", &Object::type, "Object type");
+    object.def("getName", &Object::getObjectName, "Object name");
+    object.def("__repr__", &Object::print, "String representation of the object", py::arg("prepend")=std::string(""));
+    object.def_readwrite("namespace", &Object::ns_);
+    object.def_readwrite("debugMode", &Object::debug_);
+
+    py::enum_<RotationType>(module, "RotationType")
+        .value("Quaternion", RotationType::QUATERNION)
+        .value("RPY", RotationType::RPY)
+        .value("ZYZ", RotationType::ZYZ)
+        .value("ZYX", RotationType::ZYX)
+        .value("AngleAxis", RotationType::ANGLE_AXIS)
+        .value("Matrix", RotationType::MATRIX)
+        .export_values();
+
+    py::class_<TaskMap, std::shared_ptr<TaskMap>, Object> taskMap(module, "TaskMap");
+    taskMap.def_readonly("id", &TaskMap::Id);
+    taskMap.def_readonly("start", &TaskMap::Start);
+    taskMap.def_readonly("length", &TaskMap::Length);
+    taskMap.def_readonly("startJ", &TaskMap::StartJ);
+    taskMap.def_readonly("lengthJ", &TaskMap::LengthJ);
+    taskMap.def("getFrames",&TaskMap::GetFrames, py::return_value_policy::copy);
+    taskMap.def("taskSpaceDim", (int (TaskMap::*)()) &TaskMap::taskSpaceDim);
+    taskMap.def("taskSpaceJacobianDim", &TaskMap::taskSpaceJacobianDim);
+    taskMap.def("debug", &TaskMap::debug);
+
+    py::class_<TaskSpaceVector, std::shared_ptr<TaskSpaceVector>> taskSpaceVector(module, "TaskSpaceVector");
+    taskSpaceVector.def("setZero", &TaskSpaceVector::setZero);
+    taskSpaceVector.def_readwrite("data", &TaskSpaceVector::data);
+    taskSpaceVector.def("__sub__", &TaskSpaceVector::operator-, py::is_operator());
+    taskSpaceVector.def("__repr__", [](TaskSpaceVector* instance){return ((std::ostringstream&)(std::ostringstream("")<<"TaskSpaceVector ("<<instance->data.transpose()<<")")).str();});
+
+    py::class_<MotionSolver, std::shared_ptr<MotionSolver>, Object> motionSolver(module, "MotionSolver");
+    motionSolver.def("specifyProblem", &MotionSolver::specifyProblem, "Assign problem to the solver", py::arg("planningProblem"));
+    motionSolver.def("solve", [](std::shared_ptr<MotionSolver> sol, Eigen::VectorXd q0){return Solve(sol, q0);}, "Solve the problem", py::arg("q0"));
+    motionSolver.def("getProblem", &MotionSolver::getProblem, py::return_value_policy::reference_internal);
+
+    py::class_<PlanningProblem, std::shared_ptr<PlanningProblem>, Object> planningProblem(module, "PlanningProblem");
+    planningProblem.def("getTasks", &PlanningProblem::getTasks, py::return_value_policy::reference_internal);
+    planningProblem.def("getScene", &PlanningProblem::getScene, py::return_value_policy::reference_internal);
+    planningProblem.def("__repr__", &PlanningProblem::print, "String representation of the object", py::arg("prepend")=std::string(""));
+
+    // Problem types
+    py::module prob = module.def_submodule("Problems","Problem types");
```

Comment:
Only code in the ```exotica``` package has to be added and only when we want it exposed.
Sub-modules are only for creating nicer looking hierarchies within python.
Other packages can create their own modules or they can add classes tho the main module. This is supported dynamically [across different libraries](http://pybind11.readthedocs.io/en/stable/advanced/misc.html#partitioning-code-over-multiple-extension-modules).